### ISSUE TITLE
BF16 PJT overview

### ISSUE THREAD

**Issue by heeju-kim2** [heeju-kim2/workspace_w-intern#3]
#### Goal
- 다양한 모델 / task / Parameter Efficient Fine-tuning에 대하여 BF16 training 시 성능 관찰 및 분석

#### Role
- 대표적인 PEFT 방법으로 4~5개 NLP/Mulit-modal task에 대해서 BF16 finetuning 때 성능 비교분석

**Comment by heeju-kim2** [comment]
### Evaluation config 정리
  - Precision
      - FP32 (baseline)
      - FP32 + BF16
      - BF16 only

  - Model
      - Llama-2-7B-chat
      - Llama-3-8B-instruct
      - Whisper-large-1.54B
  - PEFT
      - LoRA
          - LoRA: Low-Rank Adaptation of Large Language Models
      - Adapter:
          - Parameter-efficient transfer learning for NLP
      - Representation
          - ReFT: Representation Finetuning for Language Models
      - Prompt tuning
      - Prefix tuning
  - Task/Dataset
      - language
          - Chat summarization : Samsum dataset
              - LoRA, Representation tuning, Prefix tuning, Prompt tuning
              - Metric : PPL, Rouge-L
          - Math reasoning: GSM8K
              - LoRA, Representation tuning, Prefix tuning, Prompt tuning
              - Metric: EM
          - Instruction tuning: alpaca dataset
              - LoRA, Representation tuning, Prefix tuning, Prompt tuning
              - Metric: alpaca_eval 2.0
          - LongBench
              - LongLoRA, LongQLoRA
              - Metic: accuracy
      - Audio & Language
          - automatic speech recognition : ESB benchmark (TBD)
              - LoRA, Representation tuning, Prefix tuning, Prompt tuning
              - Metric: word error rate (wer)

**Comment by heeju-kim2** [comment]
| 일정  | 대엽님 업무 내용                                                                                                        |
|-------|-------------------------------------------------------------------------------------------------------------------------|
| 1주차 | 환경 세팅, llama3 task1 lora finetuning 돌려보기 (bf16, mixed, fp32), 논문 발표 (금)                                    |
| 2주차 | samsum dataset에 PEFT finetuning (LoRA, Representation, Prefix, Prompt) → 성능 측정, wg 회의                            |
| 3주차 | alpaca, GSM8K 에 LoRA finetuning 및 성능 비교, Novel RAG system 조사 (ex. Generative Representation Instruction Tuning) |
| 4주차 | LongLoRA 로 LongAlpca finetuning, LongBench-E 에 성능 확인, wg 회의, 대엽님 반차(월, 목)                                |
| 5주차 | Whisper PEFT finetuning  및 성능 측정, CRAG benchmark 돌려보기, RAG system 조사 발표 (금)                               |
| 6주차 | Whisper PEFT finetuning  및 성능 측정, wg 회의 (bf16 language part 결과 정리 발표(7/31)), CRAG benchmark porting        |
| 7주차 | Whisper PEFT finetuning  및 성능 측정, novel RAG system에 대한 평가                                                     |
| 8주차 | wg 회의, 코드 정리, 팀 전체 인턴 성과 발표                                                                              |

**Comment by dylee-furiosa** [comment]
bf16 only, mixed(bf16 + fp32), fp32 only 실험 진행
[wandb링크](https://wandb.ai/big_whale/llama_recipes/workspace?nw=nwuserbig_whale)

**Comment by dylee-furiosa** [comment]
# bf16 only
- model parameter, hidden rep, logit 모두 bfloat16 으로 계산
- logit bflaot16 -> float32로 casting 후 loss float32로 계산

# mixed (bf16 + fp32)
- model parameter float32 -> bfloat 16으로 casting 후 hidden rep bfloat 16으로 계산
- logit bfloat 16 -> float32으로 casting 후 loss float32로 계산

# fp32 only
- model parameter, hidden rep, logit, loss 모두 float32

**Comment by dylee-furiosa** [comment]
# bf16 only, mixed(bf16 + fp32), fp32 only 실험 진행
[wandb링크](https://wandb.ai/big_whale/llama_recipes/workspace?nw=nwuserbig_whale)

**Comment by dylee-furiosa** [comment]
# bf16 only
        --pure_bf16 --optimizer anyprecision --mixed_precision False --use_fp16 False \


# # bf16 + fp32 mixed precision
        --mixed_precision True --use_fp16 False \


# fp32 only
        --mixed_precision False --use_fp16 False \

**Comment by dylee-furiosa** [comment]
![datatypes_page-0001](https://github.com/heeju-kim2/workspace_w-intern/assets/173657957/5c4e14f2-12db-480d-a261-5b9395cc3bde)
![datatypes_page-0002](https://github.com/heeju-kim2/workspace_w-intern/assets/173657957/f1bb0858-fd5a-4cc2-915f-d1bed2e399e2)
